java-checks/src/main/java/org/sonar/java/checks/SillyStringOperationsCheck.java

```diff
@@ -0,0 +1,183 @@
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree tree) {
+    if (tree.methodSelect().is(Kind.MEMBER_SELECT)) {
+      boolean issue;
+      ExpressionTree str = ((MemberSelectExpressionTree) tree.methodSelect()).expression();
+      Arguments args = tree.arguments();
+      String method = tree.symbol().name();
+      switch (method) {
+        case "contains":
+        case "compareTo":
+        case "compareToIgnoreCase":
+        case "endsWith":
+        case "indexOf":
+        case "lastIndexOf":
+        case "matches":
+        case "split":
+        case "startsWith":
+          issue = checkStartsWith(str, args);
+          break;
+        case "replaceAll":
+        case "replaceFirst":
+          issue = checkReplaceFirst(str, args);
+          break;
+        case "substring":
+          issue = checkSubstring(str, args);
+          break;
+        default:
+          issue = false;
+      }
+      if (issue) {
+        reportIssue(tree, format("Remove this \"%s\" call; it has predictable results.", method));
+      }
+    }
+  }
+
+  private static boolean checkStartsWith(ExpressionTree str, Arguments args) {
+    return isSameString(str, args.get(0));
+  }
+
+  private static boolean checkReplaceFirst(ExpressionTree str, Arguments args) {
+    return isSameString(str, args.get(0)) || isSameString(args.get(0), args.get(1));
+  }
+
+  private static boolean checkSubstring(ExpressionTree str, Arguments args) {
+    if (args.size() == 1) {
+      return isZero(args.get(0)) || isStringLength(str, args.get(0));
+    } else {
+      return isStringLength(str, args.get(0)) || (isZero(args.get(0)) && isStringLength(str, args.get(1)));
+    }
+  }
+
+  private static boolean isSameString(ExpressionTree str, ExpressionTree tree) {
+    return isSameSymbol(str, tree) || isSameStringLiteral(str, tree);
+  }
+
+  private static boolean isSameSymbol(ExpressionTree tree, ExpressionTree other) {
+    return Optional.ofNullable(symbol(tree)).map(s -> s.equals(symbol(other))).orElse(false);
+  }
+
+  private static boolean isSameStringLiteral(ExpressionTree str, ExpressionTree tree) {
+    return Optional.ofNullable(string(str)).map(s -> s.equals(string(tree))).orElse(false);
+  }
+
+  private static boolean isZero(ExpressionTree tree) {
+    return Optional.ofNullable(ConstantUtils.resolveAsIntConstant(tree)).map(n -> n == 0).orElse(false);
+  }
+
+  private static boolean isStringLength(ExpressionTree str, ExpressionTree tree) {
+    Symbol symbol = symbol(str);
+    if (symbol != null && tree.is(Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree invocation = (MethodInvocationTree) tree;
+      if (STRING_LENGTH.matches(invocation) && invocation.methodSelect().is(Kind.MEMBER_SELECT)) {
+        return symbol.equals(symbol(((MemberSelectExpressionTree) invocation.methodSelect()).expression()));
+      }
+    }
+    return false;
+  }
+
+  @CheckForNull
+  private static Symbol symbol(ExpressionTree tree) {
```

Comment:
You should annotate parameter with `@Nullable` , or if you don't need to handle `null` argument, you don't need `ofNullable`